File: DSAlgorithms/Leetcode/c++/old/110-balancedBinaryTree.cpp
#include <algorithm>
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
  };

bool isBalanced(TreeNode* root){
    if(!root)
        return true;
    int leftHeight = getHeight(root->left);
    int rightHeight = getHeight(root->right);
    return std::abs(leftHeight - rightHeight) <= 1 && isBalanced(root->left) && isBalanced(root->right);
}

int getHeight(TreeNode* root){
    if(!root)
        0;
    return 1 + std::max(getHeight(root->left), getHeight(root->right));
}

File: dsa_algo/leetcode/ts/src/old/110-balancedBinaryTree.ts
const isBalanced = (root: TreeNode | null): bool => {
    if(!root)
        return true;
    let leftHeight: number = getHeight(root.left);
    let rightHeight: number = getHeight(root.right);
    return Math.abs(leftHeight - rightHeight) <= 1 && isBalanced(root.left) && isBalanced(root.right);
}

const getHeight = (root: TreeNode | null): number => {
    if(!root)
        return 0;
    return 1 + Math.max(getHeight(root.left) + getHeight(root.right));
}